File: app/src/main/java/no/joharei/flixr/api/models/Photo.java
package no.joharei.flixr.api.models;

import android.os.Parcel;
import android.os.Parcelable;

import java.util.Locale;

import no.joharei.flixr.utils.Constants;

public class Photo implements Parcelable {

    public static final Creator<Photo> CREATOR = new Creator<Photo>() {
        @Override
        public Photo createFromParcel(Parcel in) {
            return new Photo(in);
        }

        @Override
        public Photo[] newArray(int size) {
            return new Photo[size];
        }
    };
    private long id;
    private String secret;
    private int server;
    private int farm;
    private String title;
    private String isprimary;
    private int ispublic;
    private int isfriend;
    private int isfamily;

    private Photo(Parcel in) {
        id = in.readLong();
        secret = in.readString();
        server = in.readInt();
        farm = in.readInt();
        title = in.readString();
        isprimary = in.readString();
        ispublic = in.readInt();
        isfriend = in.readInt();
        isfamily = in.readInt();
    }

    @Override
    public void writeToParcel(Parcel parcel, int i) {
        parcel.writeLong(id);
        parcel.writeString(secret);
        parcel.writeInt(server);
        parcel.writeInt(farm);
        parcel.writeString(title);
        parcel.writeString(isprimary);
        parcel.writeInt(ispublic);
        parcel.writeInt(isfriend);
        parcel.writeInt(isfamily);
    }

    @Override
    public int describeContents() {
        return 0;
    }

    public long getId() {
        return id;
    }

    public String getTitle() {
        return title;
    }

    public String getCardImageUrl() {
        return String.format(Locale.getDefault(),
                Constants.THUMBNAIL_URL_FORMAT,
                farm,
                server,
                id,
                secret
        );
    }

    public String getFullscreenImageUrl() {
        return String.format(Locale.getDefault(),
                Constants.FULLSCREEN_URL_FORMAT,
                farm,
                server,
                id,
                secret
        );
    }
}

File: app/src/main/java/no/joharei/flixr/api/models/Photo.kt
package no.joharei.flixr.api.models

import android.os.Parcel
import android.os.Parcelable
import com.google.gson.annotations.SerializedName
import no.joharei.flixr.common.adapters.PhotoItem
import no.joharei.flixr.common.getUrlOfSmallestPhotoToFillSize
import no.joharei.flixr.common.getUrlsOfLargerPhotos
import paperparcel.PaperParcel

@PaperParcel
data class Photo(
        val id: Long,
        val secret: String,
        val server: Int,
        val farm: Int,
        val title: String,
        @SerializedName("isprimary")
        val isPrimary: String,
        @SerializedName("ispublic")
        val isPublic: Int,
        @SerializedName("isfriend")
        val isFriend: Int,
        @SerializedName("isfamily")
        val isFamily: Int,
        val widthN: Int,
        val widthZ: Int?,
        val widthC: Int?,
        val widthB: Int?,
        val widthH: Int?,
        val widthK: Int?,
        val widthO: Int?,
        val heightN: Int,
        val heightZ: Int?,
        val heightC: Int?,
        val heightB: Int?,
        val heightH: Int?,
        val heightK: Int?,
        val heightO: Int?,
        val urlN: String?,
        val urlZ: String?,
        val urlC: String?,
        val urlB: String?,
        val urlH: String?,
        val urlK: String?,
        val urlO: String?
) : PhotoItem(), Parcelable {
    override val thumbnailWidth: Int get() = widthN
    override val thumbnailHeight: Int get() = heightN

    override fun writeToParcel(dest: Parcel?, flags: Int) {
        dest?.let { PaperParcelPhoto.writeToParcel(this, it, flags) }
    }

    override fun describeContents() = 0

    override fun photoUrl(width: Int, height: Int) = getUrlOfSmallestPhotoToFillSize(
            width,
            height,
            arrayOf(widthN, widthZ, widthC, widthB, widthH, widthK, widthO).filterNotNull(),
            arrayOf(heightN, heightZ, heightC, heightB, heightH, heightK, heightO).filterNotNull(),
            arrayOf(urlN, urlZ, urlC, urlB, urlH, urlK, urlO).filterNotNull()
    )

    fun alternativeUrls(width: Int, height: Int) = getUrlsOfLargerPhotos(
            width,
            height,
            arrayOf(widthN, widthZ, widthC, widthB, widthH, widthK, widthO).filterNotNull(),
            arrayOf(heightN, heightZ, heightC, heightB, heightH, heightK, heightO).filterNotNull(),
            arrayOf(urlN, urlZ, urlC, urlB, urlH, urlK, urlO).filterNotNull()
    )

    companion object {
        @Suppress("unused")
        @JvmField val CREATOR = PaperParcelPhoto.CREATOR
    }
}
